### PR TITLE
CVAT 3D - Milestone-4

### PR DESCRIPTION
**Pull request by manasars**
CVAT-3D-Milestone4 : Implement initial cuboid placement in 3D View and selected cuboid adjustment in Top, Side and Front views.

Changes include:
Implemented initial cuboid placement in 3D View.
Implemented display of top, front, and side view of the active cuboid.
Implemented adjusting the borders of views to allow changing the width and height of a specific view.
Implemented centre axis for PCD rotation.
Fixed the UTF-8 encoding exception which occurred while creating zip file in MACOS. 

Test Cases:
Manual Unit testing done locally.
Existing test cases work as expected.
System Test cases will be shared.

[x ] I submit my changes into the develop branch

[ x] I submit my code changes under the same MIT License that covers the project.

### CODE REVIEW

File: cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss [hunk 1]
```diff
@@ -263,3 +263,44 @@
...
+.react-resizable-handle-vertical-side {
+    position: absolute;
+    width: $grid-unit-size/2;
+    margin-right: auto;
+    top: 50%;
+    right: 0;
+    bottom: 0;
+    background-color: grey;
+    height: $grid-unit-size * 10;
+    cursor: pointer;
```
Comment (vnishukov) [inline cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss:293]: `cursor: ew-resize`

File: cvat-canvas3d/src/typescript/canvas3dView.ts [hunk 2]
```diff
@@ -165,94 +192,237 @@ export class Canvas3dViewImpl implements Canvas3dView, Listener {
...
+            if (data.enabled && this.mode === Mode.IDLE) {
+                this.mode = Mode.DRAW;
+                this.cube = new CuboidModel();
+            } else if (this.mode !== Mode.IDLE) {
+                this.cube = new CuboidModel();
+            }
+        } else if (reason === UpdateReasons.CANCEL) {
+            if (this.mode === Mode.DRAW) {
+                this.controller.drawData.enabled = false;
+                this.views.perspective.scene.children[0].remove(this.cube.perspective);
```
Comment (vnishukov) [inline cvat-canvas3d/src/typescript/canvas3dView.ts:206]: `
Object.keys(this.views).forEach((view: string): void => view.scene.children[0].remove(this.cube[view]))
`

**Review by vnishukov — CHANGES_REQUESTED** [review]

File: cvat-ui/package.json [hunk 3]
```diff
@@ -76,7 +76,10 @@
     "react-color": "^2.19.3",
     "react-cookie": "^4.0.3",
     "react-dom": "^16.14.0",
+    "react-hotkeys": "^2.0.0",
```
Comment (bsekachev) [inline cvat-ui/package.json:79]: Please, merge current develop before. We do not use "react-hotkeys" anymore.

**Comment by bsekachev** [comment]
Hi guys, here are my comments about the patch. I will check the code additionally.

- N shortcut does not work to draw a shape
- Shift+N shortcut does not work to redraw a shape
- I cannot up/down cuboids. 

Weren't these features implemented yet?

- As I understand we decided that cuboid track is not applicable for 3D, right? Why do we still have options between a shape and a track when create a cuboid? Also we have options (from rectangle, from 4 points) which are also inapplicable here I believe. 

- Looks like keys: U I O J K L do not work. 

- Text is not fitted for side views when they are thin (screenshot below)
Screenshot from 2021-03-12 11-31-55

- I noticed strange behavior when cuboid and view are moved during I actually do not move the cursor (attached gif below, second half)
Peek 2021-03-12 11-37

- When I start drawing performance is awful (a lot of freezes, according to chrome profiler each mouse movement call 250ms of computations). I have Core i7 8th generation CPU. A screenshot also attached. Are you planning to work on it?

Screenshot from 2021-03-12 11-46-47

- Also when I open a task, in console I have some warnings:

```
Warning: Failed prop type: Invalid prop `height` of type `string` supplied to `Resizable`, expected `number`.
    in Resizable (created by ResizableBox)
    in ResizableBox (created by CanvasWrapperComponent)
    in div (created by CanvasWrapperComponent)
    in main (created by Basic)
    in Basic (created by Content)
    in Content (created by CanvasWrapperComponent)
    in CanvasWrapperComponent (created by ConnectFunction)
    in ConnectFunction (created by StandardWorkspace3DComponent)
    in section (created by BasicLayout)
    in BasicLayout (created by Layout)
    in Layout (created by StandardWorkspace3DComponent)
    in StandardWorkspace3DComponent (created by AnnotationPageComponent)
    in main (created by Basic)
    in Basic (created by Content)
    in Content (created by AnnotationPageComponent)
    in section (created by BasicLayout)
    in BasicLayout (created by Layout)
    in Layout (created by AnnotationPageComponent)
    in AnnotationPageComponent (created by ConnectFunction)
    in ConnectFunction (created by Context.Consumer)
    in withRouter(Connect(AnnotationPageComponent)) (created by Context.Consumer)
    in Route (created by CVATApplication)
    in Switch (created by CVATApplication)
    in GlobalHotKeys (created by CVATApplication)
    in main (created by Basic)
    in Basic (created by Content)
    in Content (created by CVATApplication)
    in section (created by BasicLayout)
    in BasicLayout (created by Layout)
    in Layout (created by CVATApplication)
    in GlobalErrorBoundary (created by ConnectFunction)
    in ConnectFunction (created by CVATApplication)
    in CVATApplication (created by Context.Consumer)
    in withRouter(CVATApplication) (created by ConnectFunction)
    in ConnectFunction
    in Router (created by BrowserRouter)
    in BrowserRouter
    in Provider
```

**Comment by bsekachev** [comment]
>  Implemented centre axis for PCD rotation.

I am not sure it is in the center of .pcd (gif attached). Please, correct me if I am wrong. 
Peek 2021-03-12 11-56

**Comment by bsekachev** [comment]
And also I have some issues with context image (It is not a part of this milestone, but we expect everything should be already worked there). 

We have two test archives: 
https://github.com/openvinotoolkit/cvat/blob/develop/cvat/apps/engine/tests/assets/test_pointcloud_pcd.zip
https://github.com/openvinotoolkit/cvat/blob/develop/cvat/apps/engine/tests/assets/test_velodyne_points.zip

Both of them contains ``.pcd`` and ``.png``. But context image can be opened only in the second. Could you please comment?
On the second we also see that context image overfitted its container. So, it also looks like a bug. 

Screenshot from 2021-03-12 12-01-09

File: cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss (hunk 1 above)
Comment (manasars) [reply cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss]: Resolved

File: cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss [hunk 4]
```diff
@@ -263,3 +263,44 @@
...
+.react-resizable-handle-vertical-top {
+    position: absolute;
+    width: $grid-unit-size/2;
+    margin-right: auto;
+    top: 50%;
+    right: 0;
+    bottom: 0;
+    background-color: grey;
+    height: $grid-unit-size * 10;
+    cursor: pointer;
```
[the remark this answers is not in the document]
Comment (manasars) [reply cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss]: Resolved

File: cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss [hunk 5]
```diff
@@ -263,3 +263,44 @@
...
+    position: absolute;
+    margin-left: auto;
+    width: $grid-unit-size * 10;
+    margin-right: auto;
+    right: 0;
+    bottom: 0;
+    left: 0;
+    background-color: grey;
+    height: $grid-unit-size/2;
+    cursor: pointer;
```
[the remark this answers is not in the document]
Comment (manasars) [reply cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss]: Resolved

File: cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss [hunk 6]
```diff
@@ -263,3 +263,44 @@
     text-align: center;
     vertical-align: middle;
 }
+
```
[the remark this answers is not in the document]
Comment (manasars) [reply cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss:265]: Resolved, resize functionality works for the entire line between all views.

File: cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx [hunk 7]
```diff
@@ -21,7 +26,11 @@ interface Props {
     data: string;
     annotations: any[];
     onSetupCanvas: () => void;
+
```
[the remark this answers is not in the document]
Comment (manasars) [reply cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx]: Resolved

File: cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx [hunk 8]
```diff
@@ -21,7 +26,11 @@ interface Props {
...
     getContextImage(): void;
+
+    onResetCanvas(): void;
```
[the remark this answers is not in the document]
Comment (manasars) [reply cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx:29]: Resolved

File: cvat-canvas3d/src/typescript/canvas3dView.ts (hunk 2 above)
Comment (manasars) [reply cvat-canvas3d/src/typescript/canvas3dView.ts]: Resolved

File: cvat-ui/package.json (hunk 3 above)
Comment (manasars) [reply cvat-ui/package.json]: Done

**Comment by manasars** [comment]
Hi @bsekachev ,

1. The cuboid track and Drawing method have now been removed, it was not pushed earlier since we discussed during Milestone-5 features discussion.
2.  U I O J K L  keys work with the combination of Alt, since we have conflicting keys as discussed in Milestone3.
3. The text for views has been changed to Top, Side and Front as per SOW, so the fit issue will not occur now.
4. The strange behaviour noticed while moving cuboid and scene has been fixed.
5. The performance of this particular PCD(kitchen.zip) is similar in supervise.ly as well. We also noticed that this PCD does not load properly in supervise.ly and looks like the pcd is very dense, we checked with other PCD files and performance seems to be fine, if any further tuning required we will look into it in the upcoming milestones.
6. PFA performance comparison for the kitchen.zip w.r.t supervise.ly

performance-comparision
7. The context image fit issue and test archive file context-image load issue have been fixed.
8. The short cut key Shift+N , N was missed from our end, let us know if it is fine to push the change along with next Milestone.

**Comment by bsekachev** [comment]
@manasars 

> The cuboid track and Drawing method have now been removed, it was not pushed earlier since we discussed during Milestone-5 features discussion.

But are you planning to remove them in the future? 

> U I O J K L keys work with the combination of Alt, since we have conflicting keys as discussed in Milestone3.

What about adding a tooltip because it is not clear that I should use 'Alt' together with these buttons. What is more, I see you use <Button> element to display shortcuts. IMHO buttons are assumed to do some actions, but they do not. Can we add the same actions for these button as shortcuts do? 

Screenshot from 2021-03-17 11-09-04

> The text for views has been changed to Top, Side and Front as per SOW, so the fit issue will not occur now.
The strange behaviour noticed while moving cuboid and scene has been fixed.
The context image fit issue and test archive file context-image load issue have been fixed.

Thanks, I'll check it one more time. @dvkruchinin could you please check if now you can write a tests for context image? 

> The short cut key Shift+N , N was missed from our end, let us know if it is fine to push the change along with next Milestone.

Please, discuss your plans with @nmanovic . 

> PFA performance comparison for the kitchen.zip w.r.t supervise.ly

As I see CVAT cuboid placing is 3x slower than in supervise.ly. I believe it should be accelerated. @nmanovic , what do you think?

File: cvat-ui/src/reducers/interfaces.ts [hunk 9]
```diff
@@ -584,3 +584,10 @@ export enum DimensionType {
     DIM_3D = '3d',
     DIM_2D = '2d',
 }
+
+export enum ViewType {
+    PERSPECTIVE = 'perspective',
+    TOP = 'top',
+    SIDE = 'side',
+    FRONT = 'front',
+}
```
Comment (bsekachev) [inline cvat-ui/src/reducers/interfaces.ts:593]: I think we can import it from canvas-3d module. We should not declare the type again.

File: cvat-ui/src/containers/annotation-page/standard-workspace/controls-side-bar/draw-shape-popover.tsx [hunk 10]
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2020 Intel Corporation
+// Copyright (C) 2021 Intel Corporation
```
Comment (bsekachev) [inline cvat-ui/src/containers/annotation-page/standard-workspace/controls-side-bar/draw-shape-popover.tsx:1]: 2020-2021

File: cvat-canvas3d/src/typescript/canvas3d.ts [hunk 11]
```diff
@@ -17,6 +19,9 @@ interface Canvas3d {
     mode(): Mode;
     render(): void;
     keyControls(keys: KeyboardEvent): void;
+    mouseControls(type: string, event: MouseEvent): void;
```
Comment (bsekachev) [inline cvat-canvas3d/src/typescript/canvas3d.ts:22]: ``type`` should be ``enum``, not just string because we have a set of possible values

File: cvat-canvas3d/src/typescript/canvas3dController.ts [hunk 12]
```diff
@@ -3,8 +3,10 @@
 // SPDX-License-Identifier: MIT
 
 import { Canvas3dModel, Mode } from './canvas3dModel';
+import { DrawData } from '../../../cvat-canvas/src/typescript/canvasModel';
```
Comment (bsekachev) [inline cvat-canvas3d/src/typescript/canvas3dController.ts:6]: You have ``DrawData`` in ``cvat-canvas3d/src/typescript/canvas3dModel.ts``. 
Probably this is wrong import.

File: cvat-canvas3d/src/typescript/canvas3dModel.ts [hunk 13]
```diff
@@ -39,6 +46,7 @@ export enum UpdateReasons {
 export enum Mode {
     IDLE = 'idle',
     DRAG = 'drag',
+    CREATE = 'create',
```
Comment (bsekachev) [inline cvat-canvas3d/src/typescript/canvas3dModel.ts:49]: Where do you use this Mode?

File: cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx [hunk 14]
```diff
@@ -2,14 +2,18 @@
...
+} from 'react';
 import Layout from 'antd/lib/layout/layout';
 import {
     ArrowUpOutlined, ArrowRightOutlined, ArrowLeftOutlined, ArrowDownOutlined,
 } from '@ant-design/icons';
+import { ResizableBox } from 'react-resizable';
 import { Workspace } from 'reducers/interfaces';
 import { Canvas3d } from 'cvat-canvas3d-wrapper';
 import ContextImage from '../standard3D-workspace/context-image/context-image';
+import { ViewType } from '../../../reducers/interfaces';
```
Comment (bsekachev) [inline cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx:16]: Please, use absolute import here from ``cvat-canvas3d-wrapper``

File: cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx [hunk 15]
```diff
@@ -59,19 +71,98 @@ const CanvasWrapperComponent = (props: Props): ReactElement => {
...
+                        ...orthographicViewSize,
+                        side,
+                        front,
+                    });
+                }
+            }
+        }
+    };
+
     useEffect(() => {
```
Comment (bsekachev) [inline cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx:166]: This useEffect works each component render, not only when component was first time mounted. 
Probably it is a mistake? We can use ``[]`` as dependency argument, in this case useEffect works on mount only.

**Comment by dvkruchinin** [comment]
@bsekachev 
> could you please check if now you can write a tests for context image?

Sure. I`ll check.

File: cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx [hunk 16]
```diff
@@ -153,34 +258,59 @@ const CanvasWrapperComponent = (props: Props): ReactElement => {
...
-            </div>
-            <div className='cvat-canvas3d-orthographic-views'>
-                <div className='cvat-canvas3d-orthographic-view cvat-canvas3d-topview'>
-                    <div className='cvat-canvas3d-header'>TOP VIEW</div>
-                    <div className='cvat-canvas3d-fullsize' ref={topView} />
-                </div>
-                <div className='cvat-canvas3d-orthographic-view cvat-canvas3d-sideview'>
-                    <div className='cvat-canvas3d-header'>SIDE VIEW</div>
-                    <div className='cvat-canvas3d-fullsize' ref={sideView} />
+            <ResizableBox
```
Comment (bsekachev) [inline cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx:269]: Please, resolve typescript issues here and in this file

Screenshot from 2021-03-17 11-41-07

File: cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/draw-shape-popover.tsx [hunk 17]
```diff
@@ -48,8 +50,93 @@ function DrawShapePopoverComponent(props: Props): JSX.Element {
...
+                                />
+                            </Col>
+                        </Row>
+                    )}
+                </>
+            );
+        }
+        return null;
+    };
+
```
Comment (bsekachev) [inline cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/draw-shape-popover.tsx:139]: Why do you need to change this file?

File: cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss [hunk 18]
```diff
@@ -263,3 +262,44 @@
...
+    position: absolute;
+    width: $grid-unit-size/2;
+    margin-right: auto;
+    top: $grid-unit-size * 4.5;
+    right: 0;
+    bottom: 0;
+    background-color: grey;
+    height: 100%;
+    cursor: ew-resize;
+}
```
Comment (bsekachev) [inline cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss:305]: For classes, defined in CVAT, please, use ``cvat-`` prefix.

**Comment by dvkruchinin** [comment]
Hi @manasars,
Can you please describe the rules for naming ```.pcd``` and ```.png``` files, as well as the structure of the directories in the zip archive for the correct operation of the context image?
As far as I understand I can't name files using the symbol ```_```.

File: cvat-ui/src/reducers/interfaces.ts (hunk 9 above)
Comment (manasars) [reply cvat-ui/src/reducers/interfaces.ts]: Resolved

File: cvat-canvas3d/src/typescript/canvas3d.ts (hunk 11 above)
Comment (manasars) [reply cvat-canvas3d/src/typescript/canvas3d.ts:22]: Resolved

File: cvat-canvas3d/src/typescript/canvas3dController.ts (hunk 12 above)
Comment (manasars) [reply cvat-canvas3d/src/typescript/canvas3dController.ts]: Resolved

File: cvat-canvas3d/src/typescript/canvas3dModel.ts (hunk 13 above)
Comment (manasars) [reply cvat-canvas3d/src/typescript/canvas3dModel.ts]: Removed unwanted mode, Resolved

File: cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx (hunk 14 above)
Comment (manasars) [reply cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx]: Resolved

File: cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx (hunk 15 above)
Comment (manasars) [reply cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx:168]: Resolved

File: cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx (hunk 16 above)
Comment (manasars) [reply cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx:271]: Resolved

File: cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss (hunk 18 above)
Comment (manasars) [reply cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss]: Resolved

**Comment by manasars** [comment]
Hi @bsekachev , we have resolved most of the comments mention, we shall push some minor review fixes tomorrow.

**Comment by manasars** [comment]
> Hi @manasars,
> Can you please describe the rules for naming `.pcd` and `.png` files, as well as the structure of the directories in the zip archive for the correct operation of the context image?
> As far as I understand I can't name files using the symbol `_`.

Hi @dvkruchinin,
PFB folder structure and naming formats allowed:

Folder- Structure 1
`-- root_folder
    |-- IMAGE_00
    |   `-- data
    |       `-- File1.png
	|       `-- File2.png
	|-- IMAGE_01
    |   `-- data
    |       `-- File1.png
	|       `-- File2.png
    `-- velodyne_points
        `-- data
            `-- File1.bin
            `-- File2.pcd
			
			
Folder- Structure 2
|-- root_folder
|   `-- File1.pcd
|   `-- File2.bin
`-- related_images
    `-- File1_pcd
        `-- 000001.png   //filename can be any thing
    `-- File2_pcd
        `-- 000001.png  //filename can be any thing

Folder-Structure 3
|-- root_folder
|   `-- File1.pcd
|   `-- File2.bin
|   `-- File1.png
|   `-- File2.png

Rules: 
- Allowed PCD types - ".bin", ".pcd" 
- Allowed image types - jpg, jpeg, png, gif, bmp
- Filenames should not start with special characters.
- In Folder-Structure 1, image filename should be same as pcd filename otherwise it'll be ignored.
- In Folder-Structure 2,inside related_images, The image folder name should be mapped corresponding to pcd file name following with "_pcd"

**Review by vnishukov — DISMISSED** [review]

File: cvat-canvas3d/src/typescript/canvas3d.ts (hunk 11 above)
Comment (bsekachev) [reply cvat-canvas3d/src/typescript/canvas3d.ts:24]: @manasars I still see simple ``string`` here

File: cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx [hunk 19]
```diff
@@ -35,6 +42,13 @@ const CanvasWrapperComponent = (props: Props): ReactElement => {
     const sideView = useRef<HTMLCanvasElement>(null);
     const frontView = useRef<HTMLCanvasElement>(null);
```
Comment (bsekachev) [inline cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx:43]: These refs are applied to HTMLDivElement, not HTMLCanvasElement. 
Probably need to fix these generic types to useRef<HTMLDivElement>

File: cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx (hunk 16 above)
Comment (bsekachev) [reply cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx:339]: @manasars 

Not exactly, I see some as before (with refs)

**Comment by bsekachev** [comment]
Hi @manasars 

I see CI tests were failed now. @dvkruchinin has setup some screenshot tests for the feature.
There is a test where two screenshots are compared (between and after wheeling TOP/SIDE/FRONT views).
The reason why the test failed is that both screenshots are the same (they are expected to be changed after wheeling).

Screenshot before:
topview_before_wheel

Screenshot after:
topview_after_wheel

You can find these screenshots pressing build detais and reviewing artifacts:
Screenshot from 2021-03-19 12-58-22


As you can see the both screenshot are equal and they are empty (no any points seen) what is unexpected. 
Personally I tried to create a task with this file **cvat/apps/engine/tests/assets/test_pointcloud_pcd.zip** and all the views except of perspective are empty. 

Screenshot from 2021-03-19 12-53-44

I've already [reported](https://github.com/openvinotoolkit/cvat/pull/2768#issuecomment-776645750) this bug in the previous milestone, but nobody was able to reproduce it. Some time after it disappeared itself and now appeared again. Also we can see it is reproducible on CI.

**Comment by manasars** [comment]
> Hi @manasars
> 
> I see CI tests were failed now. @dvkruchinin has setup some screenshot tests for the feature.
> There is a test where two screenshots are compared (between and after wheeling TOP/SIDE/FRONT views).
> The reason why the test failed is that both screenshots are the same (they are expected to be changed after wheeling).

Hi @bsekachev ,
I downloaded the screenshot from the build artefact and was able to see the side and front views.PFB attachment:
Screen Shot 2021-03-19 at 5 27 29 PM
The points are not very bright but are visible.
Also when we ran the cypress locally, I get as test cases as passed:
Screen Shot 2021-03-19 at 5 32 20 PM

We are not yet sure on what is causing the issue in the build, trying to replicate in other systems.

**Comment by bsekachev** [comment]
>  and was able to see the side and front views

Agree, that it is visible on CI screenshots (different .PCD is used on CI), but the object is much less visible on these views with file I've specified above when it could be noticeable on supervise.ly without any issues. I think we should start from fixing it.
Also I think black and blue are not enough contrast (supervise.ly used black/white as default). 

In CVAT current implementation visibility apparently depends on points density somehow. 
@dvkruchinin reported that increasing the density helps to pass these tests.


___ 
Probably one more issue is that the code in this patch affected somehow TOP view area and on screenshot we can't see any objects (as if it is displaced out of viewport)

Screenshot before wheel (develop):
before

Screenshot before wheel (this branch)
before_1

**Important note: These screenshots are done in headless mode;**
So, if the tests are successful for you, please, try additionally headless mode.

**Comment by dvkruchinin** [comment]
Hi @manasars ,

In addition to the message @bsekachev .
There is a feeling that the display of the point cloud on TOP does not work correctly (maybe I'm wrong). 
cvat-canvas3d-fullsize_develop
This is a point cloud in the form of a cube in the amount of 100,000 points. But for TOP VIEW, it looks like the camera is farther away than on the SIDE/FRONT. (The picture was taken on the develop branch).
```cy.get('.cvat-canvas3d-fullsize').first().screenshot('id_canvas3d_container')``` 

And as I wrote above @bsekachev Getting screenshots is different between the this branch and the develop branch.
This branch:
id_canvas3d_container_cvat-3d-m4
```cy.get('#canvas3d-container').screenshot('id_canvas3d_container')```
In this case ```id_canvas3d_container'``` is just the name of the image.

**Comment by manasars** [comment]
Hi @dvkruchinin ,

In Milestone-4 we implemented resizing of views due to which the div has changed and delta differs, hence difference in screenshots between develop and this branch.

**Comment by bsekachev** [comment]
Hi @manasars

Let me rephrase. It is expected, that the object is visible and centered in all the additional views regardless on availableness resizing feature. Also object looks smaller on top view (as @dvkruchinin  mentioned), what is probably because of wrong camera position setup. 

If you disagree, please explain your point of view.

**Comment by manasars** [comment]
> Hi @manasars
> 
> Let me rephrase. It is expected, that the object is visible and centered in all the additional views regardless on availableness resizing feature. Also object looks smaller on top view (as @dvkruchinin mentioned), what is probably because of wrong camera position setup.
> 
> If you disagree, please explain your point of view.

Hi @bsekachev ,

Yes agreed, the top view positioning needs to be adjusted, this came up as a part of adjusting the camera to support PCDs with extreme coordinates, we are working on the fix.

**Comment by bsekachev** [comment]
@manasars 

Please, merge the latest develop. We expect CI should be fixed.

File: kubernetes-templates/03_database_deployment.yml [hunk 20]
```diff
@@ -23,25 +23,25 @@ spec:
...
+                  name: cvat-postgres-secret
+                  key: POSTGRES_USER
+            - name: POSTGRES_PASSWORD
+              valueFrom:
+                secretKeyRef:
+                  name: cvat-postgres-secret
+                  key: POSTGRES_PASSWORD
           ports:
-          - containerPort: 5432
+            - containerPort: 5432
```
Comment (bsekachev) [inline kubernetes-templates/03_database_deployment.yml:44]: Have you done these changes?

File: kubernetes-templates/03_redis_deployment.yml [hunk 21]
```diff
@@ -19,11 +19,11 @@ spec:
...
-          limits:
-            cpu: "0.1"
+        - image: redis:4.0.5-alpine
+          name: cvat-redis
+          imagePullPolicy: Always
+          ports:
+            - containerPort: 6379
+          resources:
+            limits:
+              cpu: '0.1'
```
Comment (bsekachev) [inline kubernetes-templates/03_redis_deployment.yml:29]: Have you done these changes?

File: kubernetes-templates/04_cvat_backend_deployment.yml [hunk 22]
```diff
@@ -29,65 +29,65 @@ spec:
...
+              subPath: data
+            - mountPath: /home/django/keys
+              name: cvat-backend-data
+              subPath: keys
+            - mountPath: /home/django/logs
+              name: cvat-backend-data
+              subPath: logs
+            - mountPath: /home/django/models
+              name: cvat-backend-data
+              subPath: models
```
Comment (bsekachev) [inline kubernetes-templates/04_cvat_backend_deployment.yml:90]: Have you done these changes?

File: kubernetes-templates/04_cvat_frontend_deployment.yml [hunk 23]
```diff
@@ -25,5 +25,5 @@ spec:
           image: openvino/cvat_ui:v1.2.0
           imagePullPolicy: Always
           ports:
-          - containerPort: 80
+            - containerPort: 80
```
Comment (bsekachev) [inline kubernetes-templates/04_cvat_frontend_deployment.yml:28]: Please, look at `` kubernetes-templates/*``, I believe you didn't change these files.

Comment (manasars) [reply kubernetes-templates/04_cvat_frontend_deployment.yml:28]: Looks like pre-commit hook autocorrected the files, we shall revert.

**Comment by manasars** [comment]
We shall be pushing again, the kubernetes reversion has not happened yet

File: kubernetes-templates/04_cvat_backend_deployment.yml (hunk 22 above)
Comment (manasars) [reply kubernetes-templates/04_cvat_backend_deployment.yml]: Reverted code

File: kubernetes-templates/03_database_deployment.yml (hunk 20 above)
Comment (manasars) [reply kubernetes-templates/03_database_deployment.yml]: Reverted

File: kubernetes-templates/03_redis_deployment.yml (hunk 21 above)
Comment (manasars) [reply kubernetes-templates/03_redis_deployment.yml]: Reverted

**Comment by manasars** [comment]
Hi @bsekachev , we have merged with latest develop code and the CI build is successful.

**Review by bsekachev — APPROVED** [review]
